### PR TITLE
GEOMESA-3014 Compact rendering of FSDS metadata files

### PR DESCRIPTION
**Pull request by elahrvivaz**
Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala [hunk 1]
```diff
@@ -145,16 +145,20 @@ object StorageMetadata {
...
+   * @param sort sort fields, if any, as feature type attribute number
+   * @param bounds known bounds, if any, keyed by feature type attribute number
+   */
   case class StorageFile(
       name: String,
       timestamp: Long,
-      action: StorageFileAction.StorageFileAction = StorageFileAction.Append
+      action: StorageFileAction.StorageFileAction = StorageFileAction.Append,
+      sort: Seq[Int] = Seq.empty,
+      bounds: Seq[(Int, String, String)] = Seq.empty
```
Comment (jnh5y) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala:161]: Does this change the format for the metadata?  

If so, any migration or compatibility considerations?

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/package.scala [hunk 2]
```diff
@@ -55,15 +58,20 @@ package object metadata {
...
 
     def +(other: PartitionConfig): PartitionConfig = {
       require(action == PartitionAction.Add, "Can't aggregate into non-add actions")
       val ts = math.max(timestamp, other.timestamp)
-      PartitionConfig(name, action, files ++ other.files, count + other.count, envelope.merge(other.envelope), ts)
+      val env = if (envelope.isEmpty) { other.envelope } else if (other.envelope.isEmpty) { envelope } else {
+        val Seq(xmin1, ymin1, xmax1, ymax1) = envelope
+        val Seq(xmin2, ymin2, xmax2, ymax2) = other.envelope
+        Seq(math.min(xmin1, xmin2), math.min(ymin1, ymin2), math.max(xmax1, xmax2), math.max(ymax1, ymax2))
+      }
```
Comment (jnh5y) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/package.scala:73]: This logic makes sense.  Should it be here? 

Definitely non-blocking, I'm just surprised that it hasn't been needed before here or elsewhere.

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala (hunk 1 above)
Comment (jnh5y) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala:161]: Assuming this changes the stored metadata, are there unit tests covering the new data?

Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala:161]: it does, but the readers can handle both formats. if you write data with a newer version, you won't be able to read it back out. currently these are unused, but i wanted to add them as placeholders. there are unit tests checking the old schemas that cover this.

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/package.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/package.scala:73]: i don't think we store envelopes as untyped Seqs like this anywhere else... the impetus here is to make the json rendering as compact as possible.

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala (hunk 1 above)
Comment (jnh5y) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala:161]: Do we need to mention that in the upgrade guide?  (E.g. not to mix and match new and old writers?)

Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/StorageMetadata.scala:161]: oh, we actually already have it in there due to the key-value metadata change :D

**Pull request merged**